// @shadow-drive/sdk/dist/types/accounts/StorageAccountV2.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageAccountV2 = void 0;
const anchor_1 = require("@coral-xyz/anchor");
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
const programId_1 = require("../programId");
class StorageAccountV2 {
    constructor(fields) {
        this.immutable = fields.immutable;
        this.toBeDeleted = fields.toBeDeleted;
        this.deleteRequestEpoch = fields.deleteRequestEpoch;
        this.storage = fields.storage;
        this.owner1 = fields.owner1;
        this.accountCounterSeed = fields.accountCounterSeed;
        this.creationTime = fields.creationTime;
        this.creationEpoch = fields.creationEpoch;
        this.lastFeeEpoch = fields.lastFeeEpoch;
        this.identifier = fields.identifier;
    }
    static fetch(c, address) {
        return __awaiter(this, void 0, void 0, function* () {
            const info = yield c.getAccountInfo(address);
            if (info === null) {
                return null;
            }
            if (!info.owner.equals(programId_1.PROGRAM_ID)) {
                throw new Error("account doesn't belong to this program");
            }
            return this.decode(info.data);
        });
    }
    static fetchMultiple(c, addresses) {
        return __awaiter(this, void 0, void 0, function* () {
            const infos = yield c.getMultipleAccountsInfo(addresses);
            return infos.map((info) => {
                if (info === null) {
                    return null;
                }
                if (!info.owner.equals(programId_1.PROGRAM_ID)) {
                    throw new Error("account doesn't belong to this program");
                }
                return this.decode(info.data);
            });
        });
    }
    static decode(data) {
        if (!data.slice(0, 8).equals(StorageAccountV2.discriminator)) {
            throw new Error("invalid account discriminator");
        }
        const dec = StorageAccountV2.layout.decode(data.slice(8));
        return new StorageAccountV2({
            immutable: dec.immutable,
            toBeDeleted: dec.toBeDeleted,
            deleteRequestEpoch: dec.deleteRequestEpoch,
            storage: dec.storage,
            owner1: dec.owner1,
            accountCounterSeed: dec.accountCounterSeed,
            creationTime: dec.creationTime,
            creationEpoch: dec.creationEpoch,
            lastFeeEpoch: dec.lastFeeEpoch,
            identifier: dec.identifier,
        });
    }
    toJSON() {
        return {
            immutable: this.immutable,
            toBeDeleted: this.toBeDeleted,
            deleteRequestEpoch: this.deleteRequestEpoch,
            storage: this.storage.toString(),
            owner1: this.owner1.toString(),
            accountCounterSeed: this.accountCounterSeed,
            creationTime: this.creationTime,
            creationEpoch: this.creationEpoch,
            lastFeeEpoch: this.lastFeeEpoch,
            identifier: this.identifier,
        };
    }
    static fromJSON(obj) {
        return new StorageAccountV2({
            immutable: obj.immutable,
            toBeDeleted: obj.toBeDeleted,
            deleteRequestEpoch: obj.deleteRequestEpoch,
            storage: new anchor_1.BN(obj.storage),
            owner1: new anchor_1.web3.PublicKey(obj.owner1),
            accountCounterSeed: obj.accountCounterSeed,
            creationTime: obj.creationTime,
            creationEpoch: obj.creationEpoch,
            lastFeeEpoch: obj.lastFeeEpoch,
            identifier: obj.identifier,
        });
    }
}
exports.StorageAccountV2 = StorageAccountV2;
StorageAccountV2.discriminator = Buffer.from([
    133, 53, 253, 82, 212, 5, 201, 218,
]);
StorageAccountV2.layout = borsh.struct([
    borsh.bool("immutable"),
    borsh.bool("toBeDeleted"),
    borsh.u32("deleteRequestEpoch"),
    borsh.u64("storage"),
    borsh.publicKey("owner1"),
    borsh.u32("accountCounterSeed"),
    borsh.u32("creationTime"),
    borsh.u32("creationEpoch"),
    borsh.u32("lastFeeEpoch"),
    borsh.str("identifier"),
]);
